### PR TITLE
Fix API server config update "concurrent map iteration and map write" (bug 639)

### PR DESCRIPTION
**Pull request by Sindica**

### CODE REVIEW

File: staging/src/k8s.io/client-go/rest/config.go [hunk 1]
```diff
@@ -191,10 +191,10 @@ func (ag *Config) updateConfig() {
 
 	// assign new hosts
 	i := 0
-	klog.V(6).Infof("Len of new server %d; len of new config %d", numOfServers, len(ag.config))
+	klog.V(4).Infof("Len of new server %d; len of new config %d", numOfServers, len(ag.config))
 	for _, ss := range newApiServers {
 		ag.config[i].Host = fmt.Sprintf("%s://%s:%d/", ss.Ports[0].Name, ss.Addresses[0].IP, ss.Ports[0].Port)
```
Comment (chenqianfzh) [inline staging/src/k8s.io/client-go/rest/config.go:196]: minor: remove line 198 and change line 195 to "for i, ss := ...."

File: staging/src/k8s.io/client-go/apiserverupdate/update_test.go [hunk 2]
```diff
@@ -52,6 +77,70 @@ func TestSetAPIServerConfig(t *testing.T) {
...
+			go setAndReadAPIServerConfig(&wg, epMap1)
+			go setAndReadAPIServerConfig(&wg, epMap2)
+		}
+		wg.Wait()
+	}
+
+	// final test
+	readEPMap := GetAPIServerConfig()
+	assert.NotNil(t, readEPMap)
+	assert.True(t, len(readEPMap) <= 2)
```
Comment (chenqianfzh) [inline staging/src/k8s.io/client-go/apiserverupdate/update_test.go:122]: seems to me that we also need to check assert.True(t, len(readEPMap) >= 1)

File: staging/src/k8s.io/client-go/apiserverupdate/update_test.go [hunk 3]
```diff
@@ -52,6 +77,70 @@ func TestSetAPIServerConfig(t *testing.T) {
 	assert.Equal(t, "", ss2.Addresses[0].Hostname)
 }
 
+func setAndReadAPIServerConfig(wg *sync.WaitGroup, epMap map[string]v1.EndpointSubset) {
+	wg.Add(1)
+	SetAPIServerConfig(epMap)
+	readEPMap := GetAPIServerConfig()
+	for sg, ss := range readEPMap {
+		time.Sleep(10 * time.Millisecond)
+		klog.V(6).Infof("Make sure read server group %s and endpoints %v", sg, ss)
```
Comment (chenqianfzh) [inline staging/src/k8s.io/client-go/apiserverupdate/update_test.go:86]: incomplete log message: make sure what?

Comment (Sindica) [reply staging/src/k8s.io/client-go/apiserverupdate/update_test.go:86]: "read server group and endpoints"

File: staging/src/k8s.io/client-go/apiserverupdate/update_test.go (hunk 2 above)
Comment (Sindica) [reply staging/src/k8s.io/client-go/apiserverupdate/update_test.go:122]: line 123 does the same thing.

**Pull request merged**